src/audio/pipeline/pipeline-schedule.c

```diff
@@ -38,14 +38,68 @@ static void pipeline_schedule_cancel(struct pipeline *p)
 	sa_set_panic_on_delay(true);
 }
 
+static enum task_state pipeline_task_cmd(struct pipeline *p,
+					 struct sof_ipc_reply *reply)
+{
+	int err, cmd = p->trigger.cmd;
+
+	if (!p->trigger.host) {
+		p->trigger.cmd = -EINVAL;
+		return p->status == COMP_STATE_PAUSED ? SOF_TASK_STATE_COMPLETED :
+			SOF_TASK_STATE_RESCHEDULE;
+	}
+
+	err = pipeline_trigger_run(p, p->trigger.host, cmd);
+	if (err < 0) {
+		pipe_err(p, "pipeline_task(): failed to trigger components: %d", err);
+		reply->error = err;
+		err = SOF_TASK_STATE_COMPLETED;
+	} else {
+		switch (cmd) {
+		case COMP_TRIGGER_START:
+		case COMP_TRIGGER_RELEASE:
+			p->status = COMP_STATE_ACTIVE;
+			break;
+		case COMP_TRIGGER_PRE_START:
+		case COMP_TRIGGER_PRE_RELEASE:
+			p->status = COMP_STATE_PRE_ACTIVE;
+			break;
+		case COMP_TRIGGER_STOP:
+		case COMP_TRIGGER_PAUSE:
+			p->status = COMP_STATE_PAUSED;
+		}
+
+		if (err == PPL_STATUS_PATH_STOP) {
+			pipe_warn(p, "pipeline_task(): stopping for xrun");
+			err = SOF_TASK_STATE_COMPLETED;
+		} else if (p->trigger.cmd != cmd) {
+			/* PRE stage completed */
+			if (p->trigger.delay)
+				return SOF_TASK_STATE_RESCHEDULE;
+			/* No delay: the final stage has already run too */
+		} else if (p->status == COMP_STATE_PAUSED && !p->trigger.aborted) {
+			err = SOF_TASK_STATE_COMPLETED;
+		} else {
+			p->status = COMP_STATE_ACTIVE;
+			err = SOF_TASK_STATE_RESCHEDULE;
+		}
+	}
+
+	p->trigger.cmd = -EINVAL;
```

Comment:
wont this always return an error ?

Reply:
why return an error? This just indicates, that currently there's no trigger command to be processed by the task. If one of `START`, `STOP`, `PAUSE`, `RELEASE` arrives, `trigger.cmd` will be set to it and the trigger will be executed

Reply:
ok, can we have a comment here and return 0 (no error, but nothing todo ?)